### ISSUE TITLE
Controller hot-plug support

### ISSUE THREAD

**Issue by maxigaz** [dosbox-staging/dosbox-staging#946]
**Is your feature request related to a problem? Please describe.**
DOSBox Staging detects my DualShock 4 and Wireless XBox 360 Controller only if they are switched on before DOSBox Staging is launched. If any of them is switched off and on again, DOSBox Staging has to be restarted as well so that they can be used again.

**Describe the solution you'd like**
Automatically detect any game controller if it is connected via USB or Bluetooth.

**Describe alternatives you've considered**
Retroarch with DOSBox Pure (and I believe all the other Libretro cores) support this feature.

**Additional context**
From what I remember, the current behaviour is also present in vanilla DOSBox. I've only tested these different versions under Manjaro Linux x86-64.

The change would be useful in case I turn off any of the gamepads to save battery during a break or the XBox Controller switches itself off automatically (which happens after being idle for 15 minutes).

**Comment by Burrito78** [comment]
@nemo93 Do you have an idea how this could be supported?

**Comment by weirddan455** [comment]
I think this will have to wait until we refactor the mapper.  I took a look at this thinking it would be easy to add but `sdl_mapper.cpp` is kind of a disaster.

The best way to support hotplugging (IMO) is to listen for the events `SDL_JOYDEVICEADDED` and `SDL_JOYDEVICEREMOVED`.  However, there's some hack to reduce the number of times the joystick gets polled that calls `SDL_JoystickEventState(SDL_DISABLE);` which tells SDL not to send us joystick events and instead we poll manually at a set interval.  Apparently this is needed for Macs to reduce CPU usage according to a code comment.

OK, second best option is to call `SDL_NumJoysticks()` during the polling interval and see if the number of joysticks changed.  Then call `MAPPER_BindKeys()` to re-init (same as we do if the user changes their mapperfile at runtime).  I tried that and it doesn't work either.  The bindgroups don't get re-created out when that happens because `MAPPER_Destroy` doesn't get called so it's not actually a full destroy + init.

I'm afraid if I go any deeper, I'm likely to cause regressions.  This mapper is storing the same raw pointers in multiple different `std::vectors` and also in some C style arrays with manual indexing.  So making sense of all this while trying to avoid double free, use after free, and memory leaks is going to be challenging.

Best to nuke this file from orbit and re-write it from scratch which I believe @johnnovak is planning to do at some point.  Adding hotplug support is not that difficult but making sense of this mapper code is.

**Comment by Grounded0** [comment]
> Apparently this is needed for Macs to reduce CPU usage according to a code comment.

This could be from 2005 and irrelevant today.

**Comment by johnnovak** [comment]
> Best to nuke this file from orbit and re-write it from scratch which I believe @johnnovak is planning to do at some point. Adding hotplug support is not that difficult but making sense of this mapper code is.

Precisely.

**Comment by weirddan455** [comment]
> > Apparently this is needed for Macs to reduce CPU usage according to a code comment.
> 
> This could be from 2005 and irrelevant today.

Yeah, maybe.  That's not even the biggest problem though.  Even if I could receive the SDL events (with the hack removed), I have to make sense of these data structures to actually remove and add joysticks on command.

Another funny thing, the hack is behind a macro `REDUCE_JOYSTICK_POLLING` which is hard-coded to always be on.  If I  turn it off, I get a divide by 0 crash as soon as I push a joystick button :roll_eyes:

**Comment by Grounded0** [comment]
SDL currently supports hot-plugging robustly with latest releases. This is our very own mapper apparently ruining the day.

**Comment by johnnovak** [comment]
> SDL currently supports hot-plugging robustly with the latest releases. This is our very own mapper apparently ruining the day.

Most likely 😅 This is something that will need to wait until the mapper rewrite.

**Comment by arifogel** [comment]
@maxigaz I got hot-plugging working in Linux (Ubuntu 24.04) at least with changes I've published in PR: #4122
